src/schema/me/__tests__/followed_shows.test.ts

```diff
@@ -0,0 +1,85 @@
+/* eslint-disable promise/always-return */
+import { runAuthenticatedQuery } from "test/utils"
+import gql from "lib/gql"
+import cityData from "../../../schema/city/cityDataSortedByDisplayPreference.json"
+import { LOCAL_DISCOVERY_RADIUS_KM } from "../../../schema/city/constants"
+
+const stubResolver = () => Promise.resolve({ body: [], headers: {} })
+
+const generate_query = (params = `(first: 10)`) =>
+  gql`
+    {
+      me {
+        followsAndSaves {
+          shows${params} {
+            edges {
+              node {
+                id
+              }
+            }
+          }
+        }
+      }
+    }
+  `
+
+describe("returns followed shows for a user", () => {
+  let followedShowsLoader
+
+  beforeEach(() => {
+    followedShowsLoader = jest.fn(stubResolver)
+  })
+
+  it("generates a predictable URL with no parameters", async () => {
+    const query = generate_query()
+    await runAuthenticatedQuery(query, { followedShowsLoader })
+    expect(followedShowsLoader).toHaveBeenCalledWith({
+      size: 10,
+      offset: 0,
+      total_count: true,
+    })
+  })
+
+  it("generates a predictable URL with a city slug input", async () => {
+    const nyc = cityData[0]
+    expect(nyc.slug).toBe("new-york-ny-usa")
+    const query = generate_query(`(first: 10, city: "${nyc.slug}")`)
+    await runAuthenticatedQuery(query, { followedShowsLoader })
+    expect(followedShowsLoader).toHaveBeenCalledWith({
+      size: 10,
+      offset: 0,
+      total_count: true,
+      near: `${nyc.coordinates.lat},${nyc.coordinates.lng}`,
+      max_distance: LOCAL_DISCOVERY_RADIUS_KM,
+    })
+  })
+
+  it("throws an error if presented with an invalid city slug", async () => {
+    const query = generate_query(`(first: 10, city-slug: "this-is-not-a-city")`)
+    try {
+      await runAuthenticatedQuery(query, { followedShowsLoader })
+      // shouldn't get here - if we did then the error wasn't thrown.
+      expect(true).toBeFalsy()
+    } catch (e) {
+      expect(e).toBeDefined()
+    }
+  })
```

Comment:
It's great that you're testing this! Just also has a matcher for failure, too:

```ts
await expect(runAuthenticatedQuery(query, { followedShowsLoader })).rejects.toEqual(new WhateverErrorIsThrown("OK"))
```

You can also just expect `.rejects.toMatchInlineSnapshot()` or something to have Jest update the snapshot in the test file.

I learned this recently: https://github.com/artsy/emission/pull/1529/files#diff-18770d90e60e62acb19fdc863c698fcaR75 Jest is by far my favourite testing framework – if I geek out over tests, this is why 😁 Some more of the cool matchers are listed here: https://jestjs.io/docs/en/expect

Reply:
Thank you, your insights into structuring jest tests throughout this process has been really helpful!